### DISCUSSION TITLE
상태관리 라이브러리 사용 경험 및 선택 이유

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#16] [프론트 기술 질문 공간]
프로젝트에서 상태관리 라이브러리 사용 경험 및 선택 이유를 말씀해주세요

**Comment by michoball** [comment]
저는 주로 redux와 redux-toolkit을 이용하여 상태관리를 했었습니다.  
요즘 recoil, zustand 같은 라이브러리들이 있는 것으로 알지만, 처음 배울 때 context api와 redux (redux-toolkit)으로 상태관리를 배우기도 했고, 가장 근본이 되는 상태관리 툴을 써봐야 전체적으로 state의 workflow를 이해할 수 있을 것이라 생각해서 왠만한 프로젝트에서 계속 사용중에 있습니다. 

그치만 최근에 비동기 상태에 대한 전역관리는 react-query를 사용하려 하고 있습니다. react-query를 사용해보고나니 서버 data를 따로 저장하여 전역으로 관리하는 작업이 부자연스럽다고 여겨져 사용하게 되었습니다.  

아직 알아가고 있는 라이브러리라 좀 더 학습하여 능숙하게 사용하려 하고 있습니다.

**Comment by WooGie911** [comment]
저 또한 상태관리를 위해 주로 redux와 redux-toolkit을 사용했습니다.
물론 제가 진행했던 프로젝트들은 모두 규모가 작기에 redux를 사용하기 보단 context api 나 다른 라이브러리를 사용하는 것이 효율적일 것입니다. 하지만 규모가 큰 경우에는 상태를 체계적으로 관리할 수 있고, 코드의 유지 보수성을 높여주고, 작업 효율을 극대화 해준다는 장점을 경험, 연습하기 위해 redux를 주로 사용했습니다. 또한 redux-thunk라는 미들웨어를 사용해 비동기 작업을 효율적으로 했습니다.

**Comment by javascriptbiggosoo** [comment]
recoil을 주로 사용합니다. React 기본 hook인 useState를 다룰 줄 안다면 코드만 봐도 동작 법을 유추할 수 있을만큼 직관적이기 때문에 새로 들어오는 팀원들의 온보딩에도 유리하다고 생각합니다.

**Comment by skyhanull** [comment]
주로 redux-toolkit과 recoil을 위주로 사용합니다
redux-toolkit은 기존의 redux보다 코드의 양이 감소하고 따로 설치할 패키지의 양이 줄었기 때문입니다.또한 redux-toolkit은 기존의 redux의 flux(단방향)패턴을 가지고 있기에 문제가 생길 시 되돌아서 찾는 과정이 편리합니다
recoil은 제가 주로 쓰는 라이브러리가 react인데 이 react와 맞게 나온 상태관리 툴이기 때문에 사용했습니다. 특히 위의 redux보다 감소한 코드의 양과 atom하나로 전역상태를 관리 할 수 있다는 점, 또한 가져오거나 추가할 때 쓰는 메소드들이 기존의 react-hook의 모양과 비슷해서 사용하기 편했기에 사용하고 있습니다

**Comment by BaekYuri** [comment]
상태 관리 라이브러리로 redux와 비동기 처리 등 사이드 이펙트를 처리하기 위해 redux-saga를 사용했던 경험이 있습니다.
redux saga의 effect를 사용하면서 가장 마지막에 들어온 요청만 처리하는게 편리했고, 여러 api를 호출하는 등 조금 까다로운 비동기 작업을 할 때 수월하였습니다.
